scripts/train_lm.py

```diff
@@ -144,15 +144,70 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
             i += 1
     f.close()
 
-def FormatMetaparameters(metaparameters):
+def FormatMetaparameters(metaparameters, num_train_sets):
+    # store metaparameters as floats during checking and rerounding
     out = []
+    # final output metaparameters in string
+    out_param = []
+    # round all float metaparameters to 3rd decimal places 
+    round_index = 3
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
+        x = round(param, round_index)
         out.append(x)
-
-    return ','.join(out)
+    
+    size = len(metaparameters)
+    
+    # check1: whether 0.000 or 1.000 occur because of rounding
+    # after this step, no 0.000 or 1.000 occurs unless the original number equal
+    # to 0 or 1 (if this happen, the parameters are invalid and it exits 
+    # with a warning, while this case can be very rare)
+    for i in range(0, size): 
+        if out[i] in [0.0, 1.0]:
+            out[i] = RoundToProperDigit(metaparameters[i], round_index)
+            if out[i] in [0.0, 1.0]:
+                sys.exit("train_lm.py: there is a {0}. metaparameters should be in range (0, 1).".format(out[i]))
+    
+    # check2: check repeating values. if happen, then reround them until they are different
+    # if there exist same values originally, exit with a warning (this can be very rare) 
+    # after this step, there should be no any repeating values in set out
+    for i in range(0, size):
+        # Note: the first #num_train_sets weights are not needed to be checked (they can be the same)
+        if i < num_train_sets:
+            out_param.append(format(out[i]))
+        else:
+            x = out[i]
+            for j in range(i + 1, size):
+                y = out[j]
+                # we will round to the next digit when we detect repeated values
+                round_digit = round_index + 1
+                round_max_digit = max(len(str(metaparameters[i])) - 2, len(str(metaparameters[j])) -2)
```

Comment:
Good catch! Thanks. I'll modify it soon.
Ke

On Fri, Aug 19, 2016 at 1:35 AM, chris920820 <notifications@github.com>
wrote:

> In scripts/train_lm.py
> <https://github.com/danpovey/pocolm/pull/52#discussion_r75429618>:
>
> > +                sys.exit("train_lm.py: there is a {0}. metaparameters should be in range (0, 1).".format(out[i]))
> > +
> > +    # check2: check repeating values. if happen, then reround them until they are different
> > +    # if there exist same values originally, exit with a warning (this can be very rare)
> > +    # after this step, there should be no any repeating values in set out
> > +    for i in range(0, size):
> > +        # Note: the first #num_train_sets weights are not needed to be checked (they can be the same)
> > +        if i < num_train_sets:
> > +            out_param.append(format(out[i]))
> > +        else:
> > +            x = out[i]
> > +            for j in range(i + 1, size):
> > +                y = out[j]
> > +                # we will round to the next digit when we detect repeated values
> > +                round_digit = round_index + 1
> > +                round_max_digit = max(len(str(metaparameters[i])) - 2, len(str(metaparameters[j])) -2)
>
> The logic, I believe, still not completely right. For example,
> For input: metaparameters = [0.1231231,0.1231232,0.1231233,0.12312322]
> output: ‘0.1231231,0.1231232,0.1231233,0.1231232’ (the second and forth
> one find collision)
>
> Second, I think the reason, why input metaparameters =
> [0.000001,0.000002,0.000003] gives [0,0,0] is because len(str(0.00000001))
> = 1e-8 = 5
> So, the max precision will be 5 instead of 9. This is a very tricky thing
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/pocolm/pull/52/files/87ca2779cd82ad774f490c940435f54fe5c991b7#r75429618>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANVxStGZoUOaGLH3DzfOWUzL1ZK-wBMSks5qhUCmgaJpZM4Jl6Bt>
> .
>



-- 
Ke Li
Dept. of Electrical and Computer Engineering
Johns Hopkins University
Email: kli26@jhu.edu